core/trino-main/src/main/java/io/trino/operator/output/UnnestingPositionsAppender.java

```diff
@@ -96,7 +117,15 @@ public long getSizeInBytes()
 
     private void appendDictionary(IntArrayList positions, DictionaryBlock source)
     {
-        delegate.append(mapPositions(positions, source), source.getDictionary());
+        Block dictionary = source.getDictionary();
+        IntArrayList mappedPositions = mapPositions(positions, source);
+        if (dictionaryBlockBuilder.canAppend(dictionary)) {
+            dictionaryBlockBuilder.append(mappedPositions, dictionary);
+        }
+        else {
+            dictionaryBlockBuilder.flushDictionary(delegate);
+            delegate.append(mappedPositions, dictionary);
```

Comment:
Why don't we create new dictionaryBlockBuilder and start appending to that instead ?

Reply:
we would then potentially have multiple blocks being built e.g. flat and dictionary or possibly multiple dictionaries. That is not what we can handle in position appender.